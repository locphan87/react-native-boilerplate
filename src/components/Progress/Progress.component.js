// @flow
import React from 'react'
import { View } from 'react-native'
import { Slider } from 'react-native-elements'
import { propEq, map } from 'ramda'
import { compose, setDisplayName } from 'recompose'

import { Text } from '../../components/Styled'
import { renderWhen } from '../../hoc'
import { COLORS } from '../../themes'

type PercentProps = {
  start: number,
  current: number,
  end: number
}
type Props = PercentProps & {
  type?: string
}

const isSimple = propEq('type', 'simple')
const Percent = ({ start, current, end }: PercentProps) => {
  const percent = ((current - start) * 100) / (end - start)
  const content = `${percent.toFixed(2)}%`

  return <Text color={COLORS.yellow}>{content}</Text>
}
const Progress = ({ start, end, current, ...rest }: Props) => {
  const percentProps: PercentProps = map(Number, {
    start,
    current,
    end
  })

  return (
    <View>
      <Slider
        minimumValue={percentProps.start}
        value={percentProps.current}
        maximumValue={percentProps.end}
        {...rest}
      />
      <View style={{ alignItems: 'center' }}>
        <Percent {...percentProps} />
      </View>
    </View>
  )
}

const enhance = compose(
  setDisplayName('Progress'),
  renderWhen([
    {
      when: isSimple,
      render: Percent
    }
  ])
)

export { Progress, Percent }
export default enhance(Progress)
