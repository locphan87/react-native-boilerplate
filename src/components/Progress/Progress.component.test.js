import { testSnapshots } from '../../utils/test.util'

import { Percent, Progress } from './Progress.component'

const props = {
  start: 1,
  current: 2,
  end: 10
}

describe('Progress', () => {
  testSnapshots(Progress, [
    {
      props,
      description: 'normal progress'
    }
  ])
  testSnapshots(Percent, [
    {
      props: {
        ...props,
        type: 'simple'
      },
      description: 'simple percent'
    }
  ])
})
