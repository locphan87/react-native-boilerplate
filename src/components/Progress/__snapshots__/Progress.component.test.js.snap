// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Progress snapshots normal progress 1`] = `
<View>
  <Slider
    animationType="timing"
    debugTouchArea={false}
    maximumTrackTintColor="#b3b3b3"
    maximumValue={10}
    minimumTrackTintColor="#3f3f3f"
    minimumValue={1}
    orientation="horizontal"
    step={0}
    thumbTintColor="red"
    thumbTouchSize={
      Object {
        "height": 40,
        "width": 40,
      }
    }
    value={2}
  />
  <View
    style={
      Object {
        "alignItems": "center",
      }
    }
  >
    <Percent
      current={2}
      end={10}
      start={1}
    />
  </View>
</View>
`;

exports[`Progress snapshots simple percent 1`] = `
<StyledText
  color="yellow"
>
  11.11%
</StyledText>
`;
